index.js

```diff
@@ -3232,6 +3232,49 @@
   S.range =
   def('range', {}, [$.Integer, $.Integer, $.Array($.Integer)], range);
 
+  //# groupBy :: (a -> a -> Boolean) -> Array a -> Array (Array a)
+  //.
+  //. Splits its array argument into an array of arrays of equal,
+  //. adjacent elements. The equality of the elements is tested with
+  //. a curried function provided by the user in the first argument.
+  //.
+  //. The behaviour of groupBy can be surprising for functions that
+  //. aren't reflexive, transitive and symmetric
+  //. (see [equivalence][] relation).
+  //.
+  //. ```javascript
+  //. > S.groupBy(S.equals, [1, 1, 2, 3, 3, 4])
+  //. [[1, 1], [2], [3, 3], [4]]
+  //.
+  //. > S.groupBy(S.equals, [1, 1, 2, 1, 1])
+  //. [[1, 1], [2], [1, 1]]
+  //.
+  //. > S.groupBy(a => b => a * b % 3 === 0, [1, 2, 3, 4, 5, 6, 7, 8, 9])
+  //. [[1], [2, 3], [4], [5, 6], [7], [8, 9]]
+  //. ```
+  function groupBy(fn, xs) {
+    if (xs.length === 0) {
+      return [];
+    }
+    var acc = [[xs[0]]];
+    for (var i = 1; i < xs.length; i += 1) {
+      var last = acc[acc.length - 1];
+      var a = last[0];
+      var b = xs[i];
+      if (fn(a)(b)) {
+        last.push(b);
+      } else {
+        acc.push([b]);
+      }
+    }
+    return acc;
+  }
```

Comment:
I like these suggestions. I had a play myself, because I enjoy this sort of thing. I think this version—inspired by the suggestions above—strikes a good balance between readability and efficiency:

```javascript
function groupBy(f, xs) {
  if (xs.length === 0) return [];
  var x0 = xs[0];         // :: a
  var active = [x0];      // :: Array a
  var result = [active];  // :: Array (Array a)
  for (var idx = 1; idx < xs.length; idx += 1) {
    var x = xs[idx];
    if (f(x0)(x)) active.push(x); else result.push(active = [x0 = x]);
  }
  return result;
}
```

`result.push(active = [x0 = x])` is a dense expression, but it reads nicely to me. We're beginning a new group, so we update `x0` (which keeps track of the first value of the active group). We assign a new singleton array to `active` (which keeps track of the active group). Finally, we append the new group to `result` (which accumulates the groups).